### PR TITLE
Opuslevel

### PR DESCRIPTION
**Pull request by olofkallander**
Calculate and set audio level on forwarded packets, when needed.

### CODE REVIEW

File: bridge/engine/ActiveMediaList.cpp [hunk 1]
```diff
@@ -977,6 +977,11 @@ void ActiveMediaList::removeFromRewriteMap(size_t endpointIdHash)
     ++_ssrcMapRevision;
 }
 
+bool ActiveMediaList::isAudioLevelNeeded() const
+{
+    return _audioParticipants.size() > _audioLastN;
```
Comment (olofkallander) [inline bridge/engine/ActiveMediaList.cpp:982]: I lot of reading of the atomic in LockFreeList but we avoid decoding opus and calculating audio level in 1-1 calls and small conferences < ~3 people.
Except for BB calls as we do not know how many are on the remote side and assume auido level is needed for correct barge-in decisions.

File: bridge/engine/AudioForwarderReceiveJob.cpp [hunk 2]
```diff
@@ -150,32 +305,19 @@ void AudioForwarderReceiveJob::run()
...
-
-            _engineMixer.mapSsrc2UserId(_ssrcContext.ssrc, c9UserId);
-        }
-
-        _activeMediaList.onNewAudioLevel(_packet->endpointIdHash,
-            audioLevel.valueOr(120),
-            isPtt.isSet() && isPtt.get());
+    if (audioLevel.isSet())
+    {
+        _activeMediaList.onNewAudioLevel(_packet->endpointIdHash, audioLevel.get(), isPtt.isSet() && isPtt.get());
```
Comment (olofkallander) [inline bridge/engine/AudioForwarderReceiveJob.cpp:320]: If we have audio level ext we can optimize on near silent packets.

File: bridge/engine/ActiveMediaList.cpp [hunk 3]
```diff
@@ -538,16 +536,13 @@ void ActiveMediaList::process(const uint64_t timestamp,
         outAudioMapChanged |= updateActiveAudioList(top.participant);
         auto const& curParticipant = _audioParticipants.find(top.participant);
 
-        if (top.score - top.noiseLevel > _activeTalkerSilenceThresholdDb)
+        if (top.score > _activeTalkerSilenceThresholdDb && activeTalkersSnapshot.count < activeTalkersSnapshot.maxSize)
```
Comment (olofkallander) [inline bridge/engine/ActiveMediaList.cpp:539]: score is recent - noise , so I am not sure what this active talker selection was menat to be. One neede to be 18dB above 2x noise level to get into list.

File: bridge/engine/ActiveMediaList.cpp [hunk 4]
```diff
@@ -521,6 +517,8 @@ void ActiveMediaList::process(const uint64_t timestamp,
     size_t speakerCount = rankSpeakers();
     if (speakerCount == 0)
     {
+        TActiveTalkersSnapshot activeTalkersSnapshot;
```
Comment (olofkallander) [inline bridge/engine/ActiveMediaList.cpp:520]: If we do not set here we will have a list with at least one active talker though all are muted.

File: bridge/engine/AudioForwarderReceiveJob.cpp [hunk 5]
```diff
@@ -188,34 +330,43 @@ void AudioForwarderReceiveJob::run()
...
-        logger::error("Failed to unprotect srtp %u, mixer %s",
-            "AudioForwarderReceiveJob",
-            _ssrcContext.ssrc,
-            _engineMixer.getLoggableId().c_str());
-        return;
+        calculatedAudioLevel = decodeOpus(*_packet, !audioLevel.isSet());
+    }
+    else if (_ssrcContext.rtpMap.format == bridge::RtpMap::Format::OPUS && _needAudioLevel && !audioLevel.isSet())
+    {
+        calculatedAudioLevel = computeOpusAudioLevel(*_packet);
```
Comment (olofkallander) [inline bridge/engine/AudioForwarderReceiveJob.cpp:345]: If we have to make this feature configurable i guess we need to do something here to emulate the audio level on ptt.

**Review by RicardoMDomingues — APPROVED** [review]

**Pull request merged**